File: src/frontend/financeiro/lib/app/data/model/fin_configuracao_boleto_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:financeiro/app/data/model/model_imports.dart';
import 'package:financeiro/app/data/domain/domain_imports.dart';

class FinConfiguracaoBoletoModel {
	int? id;
	int? idBancoContaCaixa;
	String? instrucao01;
	String? instrucao02;
	String? caminhoArquivoRemessa;
	String? caminhoArquivoRetorno;
	String? caminhoArquivoLogotipo;
	String? caminhoArquivoPdf;
	String? mensagem;
	String? localPagamento;
	String? layoutRemessa;
	String? aceite;
	String? especie;
	String? carteira;
	String? codigoConvenio;
	String? codigoCedente;
	double? taxaMulta;
	double? taxaJuro;
	int? diasProtesto;
	String? nossoNumeroAnterior;
	BancoContaCaixaModel? bancoContaCaixaModel;

	FinConfiguracaoBoletoModel({
		this.id,
		this.idBancoContaCaixa,
		this.instrucao01,
		this.instrucao02,
		this.caminhoArquivoRemessa,
		this.caminhoArquivoRetorno,
		this.caminhoArquivoLogotipo,
		this.caminhoArquivoPdf,
		this.mensagem,
		this.localPagamento,
		this.layoutRemessa,
		this.aceite,
		this.especie,
		this.carteira,
		this.codigoConvenio,
		this.codigoCedente,
		this.taxaMulta,
		this.taxaJuro,
		this.diasProtesto,
		this.nossoNumeroAnterior,
		this.bancoContaCaixaModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'instrucao01',
		'instrucao02',
		'caminho_arquivo_remessa',
		'caminho_arquivo_retorno',
		'caminho_arquivo_logotipo',
		'caminho_arquivo_pdf',
		'mensagem',
		'local_pagamento',
		'layout_remessa',
		'aceite',
		'especie',
		'carteira',
		'codigo_convenio',
		'codigo_cedente',
		'taxa_multa',
		'taxa_juro',
		'dias_protesto',
		'nosso_numero_anterior',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Instrucao01',
		'Instrucao02',
		'Caminho Arquivo Remessa',
		'Caminho Arquivo Retorno',
		'Caminho Arquivo Logotipo',
		'Caminho Arquivo Pdf',
		'Mensagem',
		'Local Pagamento',
		'Layout Remessa',
		'Aceite',
		'Especie',
		'Carteira',
		'Codigo Convenio',
		'Codigo Cedente',
		'Taxa Multa',
		'Taxa Juro',
		'Dias Protesto',
		'Nosso Numero Anterior',
	];

	FinConfiguracaoBoletoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idBancoContaCaixa = jsonData['idBancoContaCaixa'];
		instrucao01 = jsonData['instrucao01'];
		instrucao02 = jsonData['instrucao02'];
		caminhoArquivoRemessa = jsonData['caminhoArquivoRemessa'];
		caminhoArquivoRetorno = jsonData['caminhoArquivoRetorno'];
		caminhoArquivoLogotipo = jsonData['caminhoArquivoLogotipo'];
		caminhoArquivoPdf = jsonData['caminhoArquivoPdf'];
		mensagem = jsonData['mensagem'];
		localPagamento = jsonData['localPagamento'];
		layoutRemessa = FinConfiguracaoBoletoDomain.getLayoutRemessa(jsonData['layoutRemessa']);
		aceite = FinConfiguracaoBoletoDomain.getAceite(jsonData['aceite']);
		especie = FinConfiguracaoBoletoDomain.getEspecie(jsonData['especie']);
		carteira = jsonData['carteira'];
		codigoConvenio = jsonData['codigoConvenio'];
		codigoCedente = jsonData['codigoCedente'];
		taxaMulta = jsonData['taxaMulta']?.toDouble();
		taxaJuro = jsonData['taxaJuro']?.toDouble();
		diasProtesto = jsonData['diasProtesto'];
		nossoNumeroAnterior = jsonData['nossoNumeroAnterior'];
		bancoContaCaixaModel = jsonData['bancoContaCaixaModel'] == null ? BancoContaCaixaModel() : BancoContaCaixaModel.fromJson(jsonData['bancoContaCaixaModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idBancoContaCaixa'] = idBancoContaCaixa != 0 ? idBancoContaCaixa : null;
		jsonData['instrucao01'] = instrucao01;
		jsonData['instrucao02'] = instrucao02;
		jsonData['caminhoArquivoRemessa'] = caminhoArquivoRemessa;
		jsonData['caminhoArquivoRetorno'] = caminhoArquivoRetorno;
		jsonData['caminhoArquivoLogotipo'] = caminhoArquivoLogotipo;
		jsonData['caminhoArquivoPdf'] = caminhoArquivoPdf;
		jsonData['mensagem'] = mensagem;
		jsonData['localPagamento'] = localPagamento;
		jsonData['layoutRemessa'] = FinConfiguracaoBoletoDomain.setLayoutRemessa(layoutRemessa);
		jsonData['aceite'] = FinConfiguracaoBoletoDomain.setAceite(aceite);
		jsonData['especie'] = FinConfiguracaoBoletoDomain.setEspecie(especie);
		jsonData['carteira'] = carteira;
		jsonData['codigoConvenio'] = codigoConvenio;
		jsonData['codigoCedente'] = codigoCedente;
		jsonData['taxaMulta'] = taxaMulta;
		jsonData['taxaJuro'] = taxaJuro;
		jsonData['diasProtesto'] = diasProtesto;
		jsonData['nossoNumeroAnterior'] = nossoNumeroAnterior;
		jsonData['bancoContaCaixaModel'] = bancoContaCaixaModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idBancoContaCaixa = plutoRow.cells['idBancoContaCaixa']?.value;
		instrucao01 = plutoRow.cells['instrucao01']?.value;
		instrucao02 = plutoRow.cells['instrucao02']?.value;
		caminhoArquivoRemessa = plutoRow.cells['caminhoArquivoRemessa']?.value;
		caminhoArquivoRetorno = plutoRow.cells['caminhoArquivoRetorno']?.value;
		caminhoArquivoLogotipo = plutoRow.cells['caminhoArquivoLogotipo']?.value;
		caminhoArquivoPdf = plutoRow.cells['caminhoArquivoPdf']?.value;
		mensagem = plutoRow.cells['mensagem']?.value;
		localPagamento = plutoRow.cells['localPagamento']?.value;
		layoutRemessa = plutoRow.cells['layoutRemessa']?.value != '' ? plutoRow.cells['layoutRemessa']?.value : 'CNAB 240';
		aceite = plutoRow.cells['aceite']?.value != '' ? plutoRow.cells['aceite']?.value : 'S';
		especie = plutoRow.cells['especie']?.value != '' ? plutoRow.cells['especie']?.value : 'DM-Duplicata Mercantil';
		carteira = plutoRow.cells['carteira']?.value;
		codigoConvenio = plutoRow.cells['codigoConvenio']?.value;
		codigoCedente = plutoRow.cells['codigoCedente']?.value;
		taxaMulta = plutoRow.cells['taxaMulta']?.value?.toDouble();
		taxaJuro = plutoRow.cells['taxaJuro']?.value?.toDouble();
		diasProtesto = plutoRow.cells['diasProtesto']?.value;
		nossoNumeroAnterior = plutoRow.cells['nossoNumeroAnterior']?.value;
		bancoContaCaixaModel = BancoContaCaixaModel();
		bancoContaCaixaModel?.nome = plutoRow.cells['bancoContaCaixaModel']?.value;
	}	

	FinConfiguracaoBoletoModel clone() {
		return FinConfiguracaoBoletoModel(
			id: id,
			idBancoContaCaixa: idBancoContaCaixa,
			instrucao01: instrucao01,
			instrucao02: instrucao02,
			caminhoArquivoRemessa: caminhoArquivoRemessa,
			caminhoArquivoRetorno: caminhoArquivoRetorno,
			caminhoArquivoLogotipo: caminhoArquivoLogotipo,
			caminhoArquivoPdf: caminhoArquivoPdf,
			mensagem: mensagem,
			localPagamento: localPagamento,
			layoutRemessa: layoutRemessa,
			aceite: aceite,
			especie: especie,
			carteira: carteira,
			codigoConvenio: codigoConvenio,
			codigoCedente: codigoCedente,
			taxaMulta: taxaMulta,
			taxaJuro: taxaJuro,
			diasProtesto: diasProtesto,
			nossoNumeroAnterior: nossoNumeroAnterior,
		);			
	}

	
}

File: backend/financeiro/src/model/fin_configuracao_boleto_model.py
from src import db
from src.model.banco_conta_caixa_model import BancoContaCaixaModel


class FinConfiguracaoBoletoModel(db.Model):
    __tablename__ = 'fin_configuracao_boleto'

    id = db.Column(db.Integer, primary_key=True)
    instrucao01 = db.Column(db.String(100))
    instrucao02 = db.Column(db.String(100))
    caminho_arquivo_remessa = db.Column(db.String(250))
    caminho_arquivo_retorno = db.Column(db.String(250))
    caminho_arquivo_logotipo = db.Column(db.String(250))
    caminho_arquivo_pdf = db.Column(db.String(250))
    mensagem = db.Column(db.String(250))
    local_pagamento = db.Column(db.String(100))
    layout_remessa = db.Column(db.String(1))
    aceite = db.Column(db.String(1))
    especie = db.Column(db.String(1))
    carteira = db.Column(db.String(3))
    codigo_convenio = db.Column(db.String(20))
    codigo_cedente = db.Column(db.String(20))
    taxa_multa = db.Column(db.Float)
    taxa_juro = db.Column(db.Float)
    dias_protesto = db.Column(db.Integer)
    nosso_numero_anterior = db.Column(db.String(50))
    id_banco_conta_caixa = db.Column(db.Integer, db.ForeignKey('banco_conta_caixa.id'))

    banco_conta_caixa_model = db.relationship('BancoContaCaixaModel', foreign_keys=[id_banco_conta_caixa])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_banco_conta_caixa = data.get('idBancoContaCaixa')
        self.instrucao01 = data.get('instrucao01')
        self.instrucao02 = data.get('instrucao02')
        self.caminho_arquivo_remessa = data.get('caminhoArquivoRemessa')
        self.caminho_arquivo_retorno = data.get('caminhoArquivoRetorno')
        self.caminho_arquivo_logotipo = data.get('caminhoArquivoLogotipo')
        self.caminho_arquivo_pdf = data.get('caminhoArquivoPdf')
        self.mensagem = data.get('mensagem')
        self.local_pagamento = data.get('localPagamento')
        self.layout_remessa = data.get('layoutRemessa')
        self.aceite = data.get('aceite')
        self.especie = data.get('especie')
        self.carteira = data.get('carteira')
        self.codigo_convenio = data.get('codigoConvenio')
        self.codigo_cedente = data.get('codigoCedente')
        self.taxa_multa = data.get('taxaMulta')
        self.taxa_juro = data.get('taxaJuro')
        self.dias_protesto = data.get('diasProtesto')
        self.nosso_numero_anterior = data.get('nossoNumeroAnterior')

    def serialize(self):
        return {
            'id': self.id,
            'idBancoContaCaixa': self.id_banco_conta_caixa,
            'instrucao01': self.instrucao01,
            'instrucao02': self.instrucao02,
            'caminhoArquivoRemessa': self.caminho_arquivo_remessa,
            'caminhoArquivoRetorno': self.caminho_arquivo_retorno,
            'caminhoArquivoLogotipo': self.caminho_arquivo_logotipo,
            'caminhoArquivoPdf': self.caminho_arquivo_pdf,
            'mensagem': self.mensagem,
            'localPagamento': self.local_pagamento,
            'layoutRemessa': self.layout_remessa,
            'aceite': self.aceite,
            'especie': self.especie,
            'carteira': self.carteira,
            'codigoConvenio': self.codigo_convenio,
            'codigoCedente': self.codigo_cedente,
            'taxaMulta': self.taxa_multa,
            'taxaJuro': self.taxa_juro,
            'diasProtesto': self.dias_protesto,
            'nossoNumeroAnterior': self.nosso_numero_anterior,
            'bancoContaCaixaModel': self.banco_conta_caixa_model.serialize() if self.banco_conta_caixa_model else None,
        }